### ISSUE TITLE
Feature: Investigate exposing a language service

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#5510]
## Overview

A TypeScript language _service_ (not to be confused a language _server_) can in theory be instantiated with `ts.createLanguageService`. As of TS 4.8 the only required parameter for that API is a `LanguageServiceHost`. If we remove the optional methods on it, it looks roughly like:

```ts
interface LanguageServiceHost {
    fileExists(path: string): boolean;
    getCompilationSettings(): ts.CompilerOptions;
    getCurrentDirectory(): string;
    getDefaultLibFileName(options: ts.CompilerOptions): string;
    getScriptFileNames(): string[];
    getScriptSnapshot(fileName: string): IScriptSnapshot | undefined;
    getScriptVersion(fileName: string): string;
}
```

Language services have APIs such as `findReferences` and `getDefinitionAtPosition` that can be quite useful for lint rules. For example, `findReferences` could quickly enable a lint rule that detects 

### Pending Questions

* How can we make sure the language service host is using the same source files & type system caching as the `ts.Program`s used for linting a file?
* What is the performance like?
* Are there any consumers of typescript-eslint that have done this thing, and what lessons have they learned?
    * Or, other TypeScript libraries? Example: [TypeStat](https://github.com/JoshuaKGoldberg/TypeStat) (referenced in #295)
* Should this be added to `createDefaultProgram`, or given opt-in as its own API?
* Per https://github.com/microsoft/TypeScript/wiki/Using-the-Language-Service-API, these are meant to be long-lived. Is there a better API we should be using?

**Comment by bradzacher** [comment]
I've investigated this before when building out https://github.com/typescript-eslint/typescript-eslint/pull/1905.

The big issue I've found is that the methods provided are too slow for linting purposes and there's a lack of caching etc.

For example something like find all references takes some 80ms per file
80ms doesn't sound like much - until you do the math on an entire lint run. 10,000 files = 800s! That's not even _that_ large of a codebase and already the lint run is dead in the water.

Most rules will take sub-1ms to run per file, at the maximum you wouldn't want a rule to take much more than 10ms per file.

I think really you would want to build new versions of the language service methods with better caching and such. There's probably also logic that could be dropped from the algorithms to speed things up.
